components/navigable-container/modes.js

```diff
@@ -0,0 +1,156 @@
+/**
+ * External Dependencies
+ */
+import { pick } from 'lodash';
+
+/**
+ * WordPress Dependencies
+ */
+import { keycodes } from '@wordpress/utils';
+
+/**
+ * Module Constants
+ */
+const { UP, DOWN, LEFT, RIGHT, TAB } = keycodes;
+
+const arrowKeys = [ UP, DOWN, LEFT, RIGHT ];
+
+function cycleValue( value, total, offset ) {
+	const nextValue = value + offset;
+	if ( nextValue < 0 ) {
+		return total + nextValue;
+	} else if ( nextValue >= total ) {
+		return nextValue - total;
+	}
+
+	return nextValue;
+}
+
+export function calculateMode( navigation ) {
+	switch ( navigation.mode ) {
+		case 'tabbing':
+			const tabConfig = { cycle: true, deep: false, widget: true, ...navigation };
+			return {
+				...pick( tabConfig, [ 'deep', 'widget', 'initialSelector' ] ),
+				useTabstops: true,
+				detect: ( event ) => {
```

Comment:
I'm not a big fan of this "mode" abstraction, Do you think we could remove it entirely in favor of something like:

```
onKeyDown( event ) {
   const offset = eventToOffset( event ) // check if we use tabs or arrows and returns 1 or -1 or null
   if ( ! offset ) {
     return;
   } 
   const tabbables = getTabbables() // depending on the deep prop
   const currentIndex = tabbables.indexOf( document.activeElement)
   const nextIndex = cycle( current, tabbables.lenght, offset ) // apply the cycle prop if necessary
   tabbables[ nextIndex ].focus();
}
```

I'm finding that the "mode" abstraction adds indirection where it's not necessary.

Reply:
Yeah, this will not support any kind of composition of keyboard navigation, and will not allow grid navigation. Maybe grid navigation isn't as important as I thought it was, but it certainly 'feels' nicer. My recommendation is to play with the InsertMenu PR (#3281) and see what I'm trying to do as a whole. That will give you a better feel about whether this level of indirection is worthwhile. The answer may definitely be "no, not really" :)

Reply:
Why can't we have something like this while keeping the `widget` prop? As I understand it, it applies when trying to get the currentIndex, so instead of `const currentIndex = tabbables.indexOf( document.activeElement)` we check the prop and use `contains` if it's a widget?

Also, I wonder if we should rename the prop. `widget` might not be explicit enough, but It's hard to find a good name. maybe `navigateContainers` (I don't really know, just thinking out loud).

---

Not saying the current implementation is bad, it's not, it's great but you know, I like simple functions ;). I find that if we can get this, it makes it clear, how each prop changes the behavior instead of having to dig to `calculateMode` and then get back to the component to understand.

Reply:
Hmm. Let's explore this.

So mode is responsible for these details:

`deep`: used by navigable container
`widget`: used by navigable container
`initialSelector` used by navigable container, but as you said, we can just make it aria-selected until it is required and drop this altogether
`useTabstops` used by navigable container to identify if looking for a tabindex of 0, or any tabindex
`detect`: used by navigable container to determine the next index based on keycodes

Note: doing this has shown me that I've got the wrong shape for the fallback return. `detect` used to be an array of rules :). I should now return a detect: () => null rather than rules: [ ].

Most of these values (cycle, deep etc.) are passed through as is (though some might have different default values). One minor part that mode is providing is working out whether the focusable targets have a tabindex of -1 (programmatic focus only) or 0 (normal tab focus) via the poorly named `useTabstops` prop. The major part that mode is providing is the `detect` function. The `detect` function has two primary roles:

a. Works out which keys are listened to. This is important. If you don't listen to a key (or stop it), then you enable your ancestors to listen to it, which can help with composing keyboard navigation. E.g. I may not care about tab, but my ancestor might.

b. Works out how to change the index based on which key was pressed. This is particularly important if you want `down` to have a different meaning from `right`. Or allow `home` to jump straight to index `0` and `end` straight to `total - 1`. This `home` and `end` capability isn't implemented now, but is mentioned as a possible improvement in the ARIA guidelines for navigating tabs (https://www.w3.org/TR/wai-aria-practices/examples/tabs/tabs-1/tabs.html)

If you wanted to get rid of mode, AND get rid of any potential for any keycode to impact the index by more than +1 or -1, then you would be able condense this down as you suggested. You would lose future flexibility, but the code would be simpler.

If you wanted to get rid of mode, AND have the ability to have more flexible index repositioning with focus (support home, end, down, up etc) then you have two options:

a) provide the `detect` function to `NavigableContainer` as a prop, and provide useTabstops as a prop.
b) provides a list of keycodes and the index changes they make (originally how `rules` worked earlier in the PR), and provide useTabstops as a prop

Essentially, this would be recreating `modes`. In (b)'s case, it would probably create more duplication, and it would force people to constantly recreate exactly the same key rules each time (right means +1, left means -1, down means +x, up means -x, home means 0, end means length - 1), and handle cycling themselves for non-trivial cases. All modes are trying to do is encompass `common` types of navigation, so that they can be reused easily. I can see how they can be slightly harder to follow, though.

My preference is to stay with `modes`, but I can also see how `a` above (passing `detect` through) would make the code easier to read. I'd be reluctant to fall back to just changing the offset by -1 or +1 depending on whether the key was a 'backward' (left, up) key or a 'forward' (right, down) key --- but maybe my prior experience with more complex navigation requirements is not relevant to Gutenberg.

I'm not sure if we'll be able to convince each other of anything without running code, so it might just be simpler if you alter this to look the way that you want, and we can then compare the advantages and disadvantages of each approach. Does that sound productive?

Reply:
Although, since the code has been refactored since the initial attempt, the calculateMode code could certainly be done in the exported components themselves (rather than passing in an object for NavigableContainer to transform). Originally, NavigableContainer was going to be exported directly.

Reply:
I see two main issues with the `detect` function:

 - It obscures simpler props
 - Composition is made more difficult, thus we loose flexibility instead of gaining flexibility IMO which means If I have a mode, that combines "tabs", "grid" and it's not provided by `calculateMode` it's hard to do unless you pass the `detect` function like you explained, but the `detect` function is not trivial and it can't be an external API of the component IMO.

Reply:
About the `-1/1` value, it was just a simplification because I didn't think we need the grid navigation right now, but if we need it, we can always change the return value of `eventToOffset` to accommodate these usecases. (`{ x: 1, y: 2}`)

Reply:
That won't handle `home` and `end` jumping to the start and end respectively. But if you're worried about having functions in the API, and you don't want modes, then I guess there isn't any alternative.